### ISSUE TITLE
schemachanger: panic and recover errors during execution

### ISSUE THREAD

**Issue by fqazi** [cockroachdb/cockroach#73604]
Currently, the declarative schema changer returns errors, which will be recovered at higher layers of the execution layer. To simplify the execution code we should instead move to a model where the execution layer panics on all unrecoverable errors, and let higher layers return these to the caller.

**Comment by postamar** [comment]
I think this is still relevant, especially but not only for `scexec`. Can we flesh this ask out a bit @fqazi @ajwerner ?

**Comment by jordanlewis** [comment]
Can we accomplish this in the short term by adding the schema change packages to `shouldCatchPanic` in `pkg/sql/colexecerror`?

**Comment by postamar** [comment]
Changing `scexec`'s error handling to a panic-and-recover scheme is doable but not trivial, however at a minimum something should be done to handle `schemachanger` panics due to runtime errors.

**Comment by jordanlewis** [comment]
Yeah I think the latter is a critical need. I don't have a comment on how important changing the whole scheme is.

**Comment by postamar** [comment]
Now that #91400 has been filed, this issue covers adopting the panic-on-error pattern in `scexec` and perhaps other packages.
